### DISCUSSION TITLE
[next] A default target for the display utility

### DISCUSSION THREAD

**Discussion by WebReflection** [pyscript/pyscript#1670] [Proposals]
## Background

With *PyScript Next* and its new core, the preferred way to have *Python code* within a *PyScript* tag is to use `<script type="py">` instead of `<py-script>`, as the latter suffers from undesired standard DOM behaviors and it gets parsed by the browser as non data-block content (it interfere with CSS parsing too and so on, as previously discussed).

Accordingly, and as it is already for PyScript "classic", scripts are smart enough to provide a companion target element **only when this is accessed** and only when something needs to be shown in place, so that having an `id` attribute for these `<script>` elements is a no-starter, as that can't even be used as target.

# Proposal

The current core offers a 100% Python `pyscript` module with some monkey-patch (around `time.sleep`, `input` or anything else that will come soon to the project + a `display` which `target` field automatically gets, when not explicitly defined, the currently executing script/tag element, so that further usage of that imported `display` will always point at the right target, still unless specified otherwise.

### Example

```html
<!doctype html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>@pyscript/core</title>
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@pyscript/core@latest/core.css" />
        <script type="module" src="https://cdn.jsdelivr.net/npm/@pyscript/core@latest/core.js"></script>
    </head>
    <body>
        <py-script id="first">
            import pyscript
            from pyscript import display
            display("Hello", append=True) # first has a DIV with Hello content
            pyscript.display("same", append=True) # first has another DIV with same
        </py-script>
        <py-script id="second">
            display("World", append=True) # first has another DIV with World
        </py-script>
        <py-script id="third">
            # we are re-importing display in here
            from pyscript import display
            display("A part", append=True)
            # third has a DIV with A part as content
            # first is not affected at all
        </py-script>
        <py-script id="fourth">
            # no matter where display is defined
            display("!", target="first", append=True)
            # first has another DIV with !
        </py-script>
    </body>
</html>
```

**Result**

![Screenshot from 2023-09-04 15-31-36](https://github.com/pyscript/pyscript/assets/85749/9d4a64de-b9a1-423f-b003-a7745434c19e)

## Implementation

While before this was implemented as *JS* module exposing a *getter* / *accessor* for the `display` property, instrumented every time through the current evaluating script, we recently moved the whole `pyscript` module into *Python* where we initially lost that handy ability to date but it recently got re-introduced by providing that very same *getter* / *accessor* within the module itself. That means that `from pyscript import display`, or even `import pyscript; pyscript.display`, will return a function that, if the `target` is not specified, uses the right target `id` out of the box available when that `import` happens.

As the already working example shows, this is completely transparent, or better, implicit, for ourselves or our users, but there are arguments against this "*magic*":

  * in *Python*, when somebody imports something from a module that is expected to be always the same
    * counter-argument: we can make `==` comparison `True` too, but `pyscript.display is pyscript.display` might return false (I say might because it depends if we want to cache the lambda per ID making this argument non existent if the executing scirpt is always the same)
  * it would be better to have an explicit utility to grab the currently executing target `id` instead to make the code more explicit and without surprises
    * counter-argument: that utility still needs to be run ASAP and outside callbacks (listener) or async tasks, because if more scripts on the page also run around that time, there can't be any guarantee that once invoked at distance that `ID` would be the expected one + there will be more boilerplate instead of less to guarantee that ID is correct
  * other ... please help us understanding your concerns, so we can update this list

## Discussion

One way to discuss this is to :+1: this post if agreeing with the current solution that works already as described in the initial example, or :-1: so that we understand what the community thinks about our current solution.

In latter case, it would be extremely nice to also explain what are your concerns or which part you really don't like, but of course that's not necessary if your thoughts are already summarized in here.

I'll start with my thoughts, please feel free to comment as much as needed/desired, thank you!

## Alternatives

Please also feel free to share alternatives to the current approach in case we've not considered these already, helping out to shape the best next release we can!

**Comment by WebReflection** [comment]
## Expectations Caveats

It's important to understand that because *PyScript Next* allows multiple scripts on the page, this issue was not really even an issue in *classic*, as more than a `<py-script>` tag had nearly no meanings because there was no way to load a different `<py-config>` or run in a *worker* and so on ... hence I'd like to clarify a few aspects of this proposal / current state:

  * with **workers**, there cannot be a default `target` whatsoever, because the code runs away from the current main thread / page. This doesn't mean we might want to provide such *default target* in the near future, it simply means this is not a concern for the *worker* use case
    * **pros** this means code in workers always need to be explicit about the `target`
    * **cons** this might be confusing for users if they expect such default target to be the script in main that asked to execute code in workers (but we can fix this if we want to, so this is less of a real-cons concern)
  * **multiple** `<script type="py">` or `<py-script>` tags on the page are not the norm, rather the exception ... *but* ... because previous version of PyScript wasn't really welcoming this approach, we expect that when multiple scripts are desired, these are handled by some centralized logic
    * **pros** there will never be an issue with the current implementation in such case if further tags don't even import `display` and assume that's been already imported, which is the *classic* expectation because `display` was global anyway
    * **cons** there will be surprises if all previous *classic* code will switch to *next* just adding `from pyscript import display` on top, as that might lead to unexpected output
  * **listeners + async** callbacks cannot reliably `from pyscript import display` at distance, neither trust a previously bound ID if that's shared globally, because `from pyscript import current_id; default_target = current_id()` works outside their scope but then such `default_target` reference would be still possibly overridden if any other script on the page does the same so that it cannot be trusted as desired one
    * **cons** no matter what we do, if we provide a way to grab a unique ID (unless we refactor a lot of code trying to be smarter) that might backfire if no different `env` is used and that retrieved *id* leaks on the global scope ... so that all it takes is to have a very good documentation around this and yet deliver the current *90% simplified use cases* out there
    * **pros** this easily gets confusing in general, even if for convoluted use cases we can always be explicit about the target, *yet* with `<script type="py">` such target might not be instantly obvious (so we need to think a way to provide it without confusing even more our users)

### Possible Solutions

These are just my personal thoughts but I sympathize with all point of views here, so that maybe I can provide some escape hatch:

  1. we drop the default `target` idea all together and we enforce users to have, or create, an element that can be used as `target` *only when showing something is desired* (beside possible terminal plugins or what's not)
  2. we instrument somehow the `pyscript` module via JS in a way that whenever anything form a specific script or tag content gets executed, including listeners, callbacks, or async stuff, with a method about which script is executing that part of the code the moment that code gets evaluated ... I think this is would be a feature request for *pyodide* project though, as I think we don't have any hook to relate executing code with the script / node / id that initially let that code evaluate /cc @hoodmane 

If I'll come up with more thoughts/alternatives, I'll surely share these.

**Reply by hoodmane** [reply]
Is there a reason we can't set the target in the setup code before executing the script? This seems to me like the most reasonable solution but maybe there are problems with it that I haven't understood.

**Reply by WebReflection** [reply]
> Is there a reason we can't set the target in the setup code before executing the script?

that means first script / tag comes, first script / tag serves every single `display` utility, so that the initial example will (???) fail expectations.

> This seems to me like the most reasonable solution but maybe there are problems with it that I haven't understood.

The example in this top-post is the reference ... your idea would fail because there is no `env` specified for the `third` script / tag. If there was an `env` the current state already works out of the box.

For workers though, the story is different ... as currently it would work out of the box if implemented, because each worker is inevitably a different environment / realm a part ... so it depends on what we would like to land, but having this consistent in workers, but not on main, might lead to extra confusion ... having this working **only in workers** though, might be an easier to explain documentation entry ... and I'd be +1 about it, it needs some refactoring but I don't see why it shouldn't work after all.

**Reply by WebReflection** [reply]
amend:

> having this consistent in workers, but not on main, might lead to extra confusion

having this consistent in workers, but bootstrapped in main as first-come first-serve for the default `display` target, might also work out of decent documentation, removing some "*hacky logic*" from our monkey-patching `stdlib/pyscript.py` file.

I would be +1 to that too.

**Reply by WebReflection** [reply]
Sorry mate, just to be sure, you +1 the discussion / proposal top-most post, I think you should've -1 that instead, if you're still concerned or prefer to discuss alternatives, which is the main goal of this discussion!

**Reply by WebReflection** [reply]
@hoodmane @antocuni @fpliger @ntoll I've created a Merge Request that addresses most concerns and gets rid of the magic behind the always different module method as accessor:

https://github.com/pyscript/pyscript/pull/1673

I think this might be ideal, less magic, still handy as it is, allowing users to still provide a `target` explicitly but doing (imho) the right thing when `display` is called.

**Comment by WebReflection** [comment]
FYI all discussions ended up agreed on this MR https://github.com/pyscript/pyscript/pull/1673

To sum it up:

  * there is no *getter* (accessor) magic involved anymore
  * there is now a `current_target()` exported utility from `pyscript` that returns the currently running/executing script/tag as *id* to use as explicit `target` in any callback or async task
  * by default, the currently executing script/tag is that target so 90% of single `<py-script>` use case are covered without any need to store/explicitly set the target`
  * with workers, the default `target` is always the script companion, when `<script type="py" worker>` is used, where such companion is created once *only* if `current_target()` is accessed, or the `<py-script worker>` itself
  * the *stdlib* has been simplified to just export different `current_target()` for either main or workers, but the `display` is a single, well defined, always same, utility

**Reply by bugzpodder** [reply]
I had a hard time getting different ids into _pyscript_js.target
pyodide.registerjsmodule is called with a static id and my run function is async so if two script are awaited together I have some trouble determining the correct Id

I had to just shim with a python version instead:
```  "_pyscript_js.py": "target = None\nPyWorker = None\n",``` 
This was what pyscript._internal.DISPLAY_TARGET is before and it worked for me so going to stick with this.

My scenario probably doesn't affect pyscript but just noting this here

**Reply by hoodmane** [reply]
Well you can still shim `current_target` in the new system.

**Reply by WebReflection** [reply]
> pyodide.registerjsmodule is called with a static id and my run function is async so if two script are awaited together I have some trouble determining the correct Id

the `current_target()` is there to solve exactly these use cases but you can always pass an explicit *target* which is the recommended way if:

  * you have more than a script in the main thread
  * your code runs asynchronously and the previous point is matched
  * your code add listeners that react at distance and the previous point is matched

If you have only one script in main you'll never have any issue with the default *target* but that's just handy for most common and basic use cases, for everything else don't use the default target and pass one explicitly or trap the `current_target()` returned ID *asap* in a way that won't conflict with other scripts in the future (i.e. by not storing it globally or it can be overwritten).

Alternatively, you can `import pyscript` and then monkey patch `pyscript.current_target` as you like but that'd be obtrusive for any other script you don't own + the default is really a handy **extra** not something you must use, I hope this makes sense.

P.S. about this part in particular

> pyodide.registerjsmodule is called with a static id

not at all, it's a getter that returns the last script/tag ID that executed the code in the main, it returns always the right thing in workers as workers are always different environments
